### PR TITLE
Upgrade to Rails 4, take 2

### LINKED ISSUE #234: Upgrade to Rails 4
With Rails 5 expected to come out this fall, development has ceased for Rails 3, and gem support has already started to deteriorate. We need to get up to speed to prevent locking ourselves into Rails 3. (it'd be even cooler if we hopped up to Rails 5 after this, but that's a stretch)

Rough order of things:
- [x] Update dependencies as much as possible & get to working state
- [x] Upgrade Rails to 4.0 with existing frameworks & backported support for dropped functionality (e.g, `attr_accessible`). Change as little code as possible.
- [ ] Convert our uses of dropped functionality (`attr_accessible`, etc) to Rails 4 conventions
- [x] Convert tests if needed
- [ ] Verify cross platform dev environments work
- [ ] Re-create staging from scratch to work with Rails 4 & deploy
- [ ] Re-create production from scratch & deploy

### PR DESCRIPTION
**Pull request by sman591**
Moving this branch into a pull request so discussion can be made per line/file. Will merge once ready.

Closes #234

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -49,12 +49,12 @@ group :assets do
...
   gem 'uglifier', '>= 1.0.3'
-end
+#end
 
 group :development do
-  gem 'better_errors'
+#  gem 'better_errors'
   gem 'binding_of_caller'
-  gem 'mail_view', '~> 2.0.4'
+ # gem 'mail_view', '~> 2.0.4'
```
Comment (sman591) [inline Gemfile]: Reason?

File: Gemfile [hunk 2]
```diff
@@ -13,7 +10,7 @@ gem 'test-unit', '~> 3.0'
 gem 'rollbar', '~> 2.7'
 gem 'skylight'
 
-gem 'sidekiq', '< 5'
+gem 'sidekiq', "~> 4"
```
Comment (sman591) [inline Gemfile]: Was there a reason for this? I think `~> 4` is the same as accepting any version

File: app/assets/stylesheets/application.css.sass [hunk 3]
```diff
@@ -1 +1 @@
-@import core
+@import "core"
```
Comment (sman591) [inline app/assets/stylesheets/application.css.sass]: Is there a need for quotes? Our other files follow the convention of no quotes

File: app/assets/stylesheets/manage.css.sass [hunk 4]
```diff
@@ -1,4 +1,4 @@
-@import core
+@import "core"
```
Comment (sman591) [inline app/assets/stylesheets/manage.css.sass]: Same here

File: app/controllers/manage/dashboard_controller.rb [hunk 5]
```diff
@@ -66,6 +69,9 @@ def schools_applied_data
 
   def activity_chart_data(types, group_type, range)
     chart_data = []
+    puts "TYPES: #{types}"
+    puts "GROUP_TYPE: #{group_type}"
+    puts "RANGE: #{range}"
```
Comment (sman591) [inline app/controllers/manage/dashboard_controller.rb]: We shouldn't be printing to the console in production

File: app/models/fips.rb [hunk 6]
```diff
@@ -1,5 +1,5 @@
 class Fips < ActiveRecord::Base
-  attr_accessible :fips_code, :city, :state
+  #attr_accessible :fips_code, :city, :state
 
-  strip_attributes
+  #strip_attributes
```
Comment (sman591) [inline app/models/fips.rb]: Why are all the `strip_attributes` calls & tests commented out?

File: app/models/school.rb [hunk 7]
```diff
@@ -1,11 +1,11 @@
 class School < ActiveRecord::Base
-  attr_accessible :name, :address, :city, :state, :bus_list_id
+  #attr_accessible :name, :address, :city, :state, :bus_list_id
```
Comment (sman591) [inline app/models/school.rb]: We can't stop using `attr_accessible` in the models without enabling [strong_parameter logic](https://github.com/rails/strong_parameters). This could cause a lot of security issues (any POST/PUT/UPDATE request can update any of the parameters otherwise, and since our source is publicized, people can know exactly what parameters to change).

File: config/application.rb [hunk 8]
```diff
@@ -1,30 +1,18 @@
...
-  Bundler.require(*Rails.groups(:assets => %w(development test)))
-  # If you want your assets lazily compiled in production, use this line
-  # Bundler.require(:default, :assets, Rails.env)
-end
+# Require the gems listed in Gemfile, including any gems
+# you've limited to :test, :development, or :production.
+Bundler.require(*Rails.groups)
 
-module BrickhackIo
+module BrickhackRail4
```
Comment (sman591) [inline config/application.rb]: Should still be BrickhackIo. Or Brickhack, but not BrickhackRail4

File: db/schema.rb [hunk 9]
```diff
@@ -54,8 +54,8 @@
     t.string   "experience"
     t.string   "interest"
     t.string   "school_id"
-    t.datetime "created_at",                                   :null => false
-    t.datetime "updated_at",                                   :null => false
+    t.datetime "created_at"
+    t.datetime "updated_at"
```
Comment (sman591) [inline db/schema.rb]: Should this be so? I feel like not allowing null for those is good? Did the convention change?

File: Gemfile [hunk 10]
```diff
@@ -27,20 +26,23 @@ gem 'roadie-rails'
 gem 'chartkick'
 gem 'groupdate'
 gem 'font-awesome-rails' # needed for icon helpers
+gem 'rails-perftest'
```
Comment (sman591) [inline Gemfile]: @gabe-k Why was this added?

File: Gemfile.lock [hunk 11]
```diff
@@ -382,6 +391,3 @@ DEPENDENCIES
   uglifier (>= 1.0.3)
   valid_attribute
   validate_url
-
-BUNDLED WITH
-   1.10.6
```
Comment (sman591) [inline Gemfile.lock]: @gabe-k What version of Bundler are you using? This line was introduced to Bundler a little while ago, and I think only old versions erase it.

File: db/schema.rb (hunk 9 above)
Comment (piedoom) [inline db/schema.rb]: I've always seen it without the `null`, unless it was specifically `null: true`.

Comment (sman591) [inline db/schema.rb]: So, the default for datetime fields should be 0 (not null), but I feel like Rails might have implemented this itself being its defined on MySQL's reference docs. Gonna check later tonight

http://dev.mysql.com/doc/refman/5.7/en/data-type-defaults.html

Comment (sman591) [inline db/schema.rb]: I did a rebuild of the `db/schema.rb` and it kept out `null: false`, so must not be needed.
